docs/parser.md

```diff
@@ -0,0 +1,247 @@
+    "field": {
+      "type": "string",
+      "pattern": "[A-Z]+"
+    }
+  }
+}
+```
+
+If we were to validate the simply JSON document `{ "field": "lowercase" }`, the validator would return output like:
+
+```
+[
+  {
+    "keyword": "pattern",
+    "dataPath": ".field",
+    "schemaPath": "#/properties/field/pattern",
+    "params": {
+      "pattern": "^[A-Z]+$"
+    },
+    "message": "should match pattern \"^[A-Z]+$\""
+  }
+]
+```
+
+Here's the same schema with the v5 error definition added.
+
+```
+{
+  "properties": {
+    "field": {
+      "type": "string",
+      "pattern": "^[A-Z]+$",
+      "errors": {
+        "pattern": "You must enter an uppercase string."
+      }
+    }
+  }
+}
+```
+
+Now the validator (with our help) can return the following output:
+
+```
+[
+  {
+    "keyword": "pattern",
+    "dataPath": ".field",
+    "schemaPath": "#/properties/field/pattern",
+    "params": {
+      "pattern": "^[A-Z]+$"
+    },
+    "message": "You must enter an uppercase string."
+  }
+]
+```
+
+## Differences from the draft v5 proposal
+
+## No support for variables
+
+The draft v5 proposal promises support for using schema variables and user data in the text of an error message. This
+module does not support this.  As such, we also  do not support the longer form of the message definition (where the
+error is an object that contains a `text` and `action`).  Instead, we support only the simplest of the forms outlined
+above, as in this updated snippet:
+
+```
+"age": {
+  "minimum": 13,
+  "errors": {
+    "minimum": "Should be at least 13 years."
+  }
+},
+```
+
+The working draft is not clear on how "required" fields should be evolved to use custom errors.  This module uses
+the schema proposed [in this GitHub issue](https://github.com/json-schema/json-schema/issues/222).  Here is an example
+of how this module expects `errors` data to be entered for required fields:
+
+```
+{
+  "properties": {
+    "shallowlyRequired": {
+      "type": "string",
+      "properties": {
+        "deeplyRequired": { "type": "string" }
+      },
+      "required": ["deeplyRequired"],
+      "errors": {
+        "required/0": "This field is required and I'm telling you about it from within a field definition."
+      }
+    }
+  },
+  "required": ["shallowlyRequired"],
+  "errors": {
+    "required/0": "This field is required and I'll tell you about it at the document level.",
+    "shallowlyRequired/required/0": "This deep field is required."
+  }
+}
+```
+
+# Requirements
+
+To use this component, you will need to instantiate it and make it aware of your schemas.  These will be dereferenced
+and cached as they are added.  You can then use the `evolveError` method to resolve the custom error definition for a
+field using the schema key and the field's JSON pointer.  See the validator implementation for an example of its use.
+
+This component uses [json-schema-ref-parser](https://github.com/BigstickCarpet/json-schema-ref-parser).  We configure
```

Comment:
This documentation should be organised as a full table of all options accepted by the component - that is, with columns, "member, type, description" as, for example, at

http://docs.fluidproject.org/infusion/development/ContextAwareness.html#-checkrecord-members-in-an-adaptationrecord-